File: app/src/main/java/cat/mvmike/minimalcalendarwidget/domain/entry/Day.java
// Copyright (c) 2016, Miquel Martí <miquelmarti111@gmail.com>
// See LICENSE for licensing information

package cat.mvmike.minimalcalendarwidget.domain.entry;

import java.text.DecimalFormat;
import java.time.DayOfWeek;
import java.time.Instant;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.time.ZoneOffset;

@SuppressWarnings("PMD.ShortClassName")
public final class Day {

    private static final String DAY_OF_MONTH_DF_PATTERN = "00";

    private final LocalDate systemLocalDate;

    private final LocalDate localDate;

    public Day(final LocalDate systemLocalDate, final LocalDate localDate) {
        this.systemLocalDate = systemLocalDate;
        this.localDate = localDate;
    }

    public boolean inYear() {
        return localDate.getYear() == systemLocalDate.getYear();
    }

    public boolean inMonth() {
        return inYear() && localDate.getMonth() == systemLocalDate.getMonth();
    }

    public boolean isToday() {
        return inMonth() && localDate.getDayOfYear() == systemLocalDate.getDayOfYear();
    }

    public DayOfWeek getDayOfWeek() {
        return localDate.getDayOfWeek();
    }

    public String getDayOfMonthString() {
        return new DecimalFormat(DAY_OF_MONTH_DF_PATTERN).format(localDate.getDayOfMonth());
    }

    public boolean isSingleDigitDay() {
        return localDate.getDayOfMonth() < 10;
    }

    public boolean isInDay(final Instant startInstant, final Instant endInstant, final boolean allDayInstance) {

        // take out 5 milliseconds to avoid erratic behaviour with full day events (or those that end at 00:00)
        return toLocalDate(startInstant, allDayInstance).getMonthValue() <= localDate.getMonthValue()
            && toLocalDate(startInstant, allDayInstance).getDayOfMonth() <= localDate.getDayOfMonth()
            && toLocalDate(endInstant.minusMillis(5), allDayInstance).getMonthValue() >= localDate.getMonthValue()
            && toLocalDate(endInstant.minusMillis(5), allDayInstance).getDayOfMonth() >= localDate.getDayOfMonth();
    }

    // calendarProvider uses different set of timezones depending if event is allDay
    private static LocalDateTime toLocalDate(final Instant instant, final boolean allDayInstance) {

        if (allDayInstance) {
            return LocalDateTime.ofInstant(instant, ZoneOffset.UTC);
        }

        return LocalDateTime.ofInstant(instant, ZoneId.systemDefault());
    }
}

File: app/src/main/kotlin/cat/mvmike/minimalcalendarwidget/domain/entry/Day.kt
// Copyright (c) 2016, Miquel Martí <miquelmarti111@gmail.com>
// See LICENSE for licensing information
package cat.mvmike.minimalcalendarwidget.domain.entry

import android.content.Context
import cat.mvmike.minimalcalendarwidget.R
import java.text.DecimalFormat
import java.time.DayOfWeek
import java.time.LocalDate

private const val DAY_OF_MONTH_DF_PATTERN: String = "00"

data class Day(
    private val systemLocalDate: LocalDate,
    val dayLocalDate: LocalDate
) {

    fun getDayOfWeek(): DayOfWeek = dayLocalDate.dayOfWeek

    fun getDayOfMonthString(): String = DecimalFormat(DAY_OF_MONTH_DF_PATTERN)
        .format(dayLocalDate.dayOfMonth.toLong())

    fun isInMonth() = dayLocalDate.year == systemLocalDate.year
            && dayLocalDate.month == systemLocalDate.month

    fun isToday() = isInMonth() && dayLocalDate.dayOfYear == systemLocalDate.dayOfYear

    fun isSingleDigitDay() = dayLocalDate.dayOfMonth < 10
}

fun getDayOfWeekDisplayValues(context: Context) =
    arrayOf(
        R.string.monday,
        R.string.tuesday,
        R.string.wednesday,
        R.string.thursday,
        R.string.friday,
        R.string.saturday,
        R.string.sunday
    ).map { dayOfWeek ->
        context.getString(dayOfWeek).replaceFirstChar { it.uppercase() }
    }.toTypedArray()

fun DayOfWeek.getAbbreviatedDisplayValue(context: Context) =
    context.getString(
        when (this) {
            DayOfWeek.MONDAY -> R.string.monday_abb
            DayOfWeek.TUESDAY -> R.string.tuesday_abb
            DayOfWeek.WEDNESDAY -> R.string.wednesday_abb
            DayOfWeek.THURSDAY -> R.string.thursday_abb
            DayOfWeek.FRIDAY -> R.string.friday_abb
            DayOfWeek.SATURDAY -> R.string.saturday_abb
            DayOfWeek.SUNDAY -> R.string.sunday_abb
        }
    ).replaceFirstChar { it.uppercase() }
